### PR TITLE
`Atlas`: Feedback loop

### PR DESCRIPTION
**Pull request by ardakaraman0**
#### General
- [ ] I chose a title conforming to the [naming conventions for pull requests](#).

#### AtlasML - API
- [ ] I updated the API documentation.

#### AtlasML - Machine Learning
- [ ] I updated the machine learning documentation.

#### Testing Instructions
Requirements:
- At least 3 Competencies with no Relations
- ...

Steps:
1. ...
2. ...

##### Code Review
- [ ] Code Review 1
- [ ] Code Review 2

##### Manuel Test
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py [hunk 1]
```diff
@@ -218,9 +216,73 @@ def newTextPipeline(self, text: str, uuid: str):
...
+                                                                  competency_id)
+        cluster_vector = np.array(found_cluster["vector"]["default"])
+        text_vector = np.array(found_text["vector"]["default"])
+
+        newMedoid = update_cluster_centroid(cluster_vector, len(members), text_vector)
+
+        new_cluster = {"properties": [{
+            "cluster_id": found_cluster["properties"]["cluster_id"],
+        }]}
+        self.weaviate_client.add_embeddings(CollectionNames.CLUSTERCENTER.value, newMedoid.tolist(), new_cluster)
```
Comment (Yhmidi) [inline atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py:288]: the feedback loop is supposedly gonna update already created clusters. Does weaviate_client.add_embeddings add if its new and update if its existing ?

File: atlas/AtlasMl/tests/test_pipelines.py [hunk 2]
```diff
@@ -114,6 +114,66 @@ def test_newTextPipeline_integration(workflows):
...
+
+    # Assert that the text now contains the competency_id in its competency_ids
+    updated_text = workflows.weaviate_client.get_embeddings_by_property("Text", "text_id", text_id)[0]
+    assert competency_id in updated_text["properties"]["competency_ids"]
+
+    # Optionally, check that the cluster has a new centroid (implementation dependent)
+    clusters = workflows.weaviate_client.get_all_embeddings("ClusterCenter")
+    found = any(cluster["id"] == cluster_id for cluster in clusters)
+    assert found, "Cluster should still exist"
+
```
Comment (Yhmidi) [inline atlas/AtlasMl/tests/test_pipelines.py:177]: I would suggest adding tests to edge cases where operation fails.

**Review by Yhmidi — CHANGES_REQUESTED** [review]

File: atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py [hunk 3]
```diff
@@ -218,9 +216,73 @@ def newTextPipeline(self, text: str, uuid: str):
...
+                                                                     competency_id)
+        found_competency = None
+        if competency: found_competency = competency[0]
+        cluster = self.weaviate_client.get_embeddings_by_property(CollectionNames.CLUSTERCENTER.value, "cluster_id",
+                                                                  found_competency["properties"]["cluster_id"])
+        found_cluster = None
+        if cluster: found_cluster = cluster[0]
+
+        new_text_competencyID = found_text["properties"]["competency_ids"]
+        new_text_competencyID.append(competency_id)
```
[the remark this answers is not in the document]
Comment (ardakaraman0) [reply atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py]: Normally the instructor can only select the competencies that are not already in the list, so I had not included that check before. Although never hurts to add more checks, addressed in https://github.com/ls1intum/edutelligence/pull/165/commits/6e57a2efd6c462807d1cfe0af4c8fd8bb8a0ea06

File: atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py (hunk 1 above)
Comment (ardakaraman0) [reply atlas/AtlasMl/atlasml/ml/MLPipelines/PipelineWorkflows.py]: Good catch thank you, addressed in https://github.com/ls1intum/edutelligence/pull/165/commits/6e57a2efd6c462807d1cfe0af4c8fd8bb8a0ea06
